### PR TITLE
Make surface tracking state a class, move functions to be methods on that class

### PR DESCRIPTION
**Pull request by peterbarker**
Also, change the type passed in to be float.

### CODE REVIEW

File: ArduCopter/mode.h [hunk 1]
```diff
@@ -17,7 +17,11 @@ class Mode {
     virtual bool init(bool ignore_checks) {
         return true;
     }
-    virtual void run() = 0;
+    void run_mode() {
+        surface_tracking.invalidate_for_logging();  // invalidate surface tracking alt, flight mode will set to true if used
```
Comment (rmackay9) [inline ArduCopter/mode.h]: I think adding another "run" is just confusing for developers.. "what's the difference between run_mode()" and run?" they will ask.  We could instead perhaps make use of an exit function or something to invalidate it or perhaps use a timer so if it's not updated for 20ms the logging stops using it.

**Comment by rmackay9** [comment]
These architectural improvements are good to see but I worry that they are dragging on a bit too long, eating up developer time that would be better spent on functional improvements.. sorry.

**Comment by peterbarker** [comment]
On Fri, 19 Apr 2019, Randy Mackay wrote:

> -    virtual void run() = 0;
> +    void run_mode() {
> +        surface_tracking.invalidate_for_logging();  // invalidate surface tracking alt, flight mode will set to true if used
> 
> I think adding another "run" is just confusing for developers.. "what's the difference between run_mode()" and run?" they will ask. We could instead
> perhaps make use of an exit function or something to invalidate it or perhaps use a timer so if it's not updated for 20ms the logging stops using it.

run_mode isn't virtual, there should be one caller to it (currently in 
update_mode, should be in the scheduler table directly).

I'd have prefered to call this method "run" and the method we call on the 
subclasses something else - but this tacking on "_mode" was the best I 
came up with to avoid pointless code churn.

File: ArduCopter/mode.h (hunk 1 above)
Comment (peterbarker) [reply ArduCopter/mode.h]: I've renamed `run_mode` to `update`.

It's unclear to me where any "exit" function would be called - we don't want to have every mode have to remember to make that call.

**Comment by rmackay9** [comment]
As discussed on the dev call, let's keep the surface tracking object to the Copter class.  Txs!

**Comment by peterbarker** [comment]
Moved this state back into Copter.

**Comment by rmackay9** [comment]
outcome of the dev call was:

- Peter to explicitely test in SITL to be sure it's working
- Randy to do line-by-line review

After this is done we should be able to merge

**Comment by peterbarker** [comment]
@rmackay9 autotest added to test this functionality.

**Comment by peterbarker** [comment]
@OXINARF - @rmackay9 is yet to look at this

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**